### ISSUE TITLE
UI Design: Drop Down Menu w/ a Responsive Design Problem (Resizing Background)

### ISSUE THREAD

**Issue by HeadfirstRYAN** [rsanchez-wsu/RaiderPlanner#232]
**Problem:** 
When the user activates the drop down menu, the content beneath it is completely blocked by the menu. This issue is designated as a UI design mishap.

![current_issue](https://user-images.githubusercontent.com/17085486/35190524-b5345688-fe31-11e7-8188-5a17d5de0fd9.PNG)
_Figure 1_

**Description:**
In _Figure 1_  the right side of the menu pane extends over the previously displayed content. The overall goal is to implement a responsive design in order for the background content to still be visible.

**Goal:** 
An important aspect in UI design is making the environment responsive. 

What would be needed is:
- [ ] Once the drop down menu is accessed for it's specific functions, the background behind the menu needs to conform around the menu.
- [ ] Once the user closes the menu, the background will reform back to it's original sizing.

**Comment by Aashishkebab** [comment]
@HeadfirstRYAN I don't think this is necessarily a good idea. Every implementation of the hamburger menu I've ever seen has the hamburger menu go over the content. Some screenshots of how hamburger menus are supposed to work. Here's Google Play Music on a desktop.
![screenshot_20180120-231638](https://user-images.githubusercontent.com/25588487/35190776-015ebbb0-fe38-11e7-8b26-0371b0951e20.png)
And then some applications on a phone.
![screenshot_20180120-231428](https://user-images.githubusercontent.com/25588487/35190769-d50437ac-fe37-11e7-8903-208c8b82137f.png)
![screenshot_20180120-231455](https://user-images.githubusercontent.com/25588487/35190770-df31a1f6-fe37-11e7-9017-f3b563df9f3b.png)

Furthermore, resizing the content could prove difficult due to the way it's rendered in the first place, but this is just speculation. More importantly, the user usually won't need to click on something in the main view because the user would have opened the hamburger menu for the sole purpose of clicking on something within it.

**Comment by Aashishkebab** [comment]
Another note. When referring to the term "responsive design", we usually talk about the application responding to a change in the size of the viewport, in this case the Window dimensions.

Last semester, I and another student were actually the ones responsible for ensuring this responsiveness. 3ad5d87 and a4ed3ed

**Comment by HeadfirstRYAN** [comment]
@Aashishkebab I'm aware of the term. I was using "responsive design" in terms of a written visual aid for the reader. This issue report is nothing if not purely stylistic. While the user would only open the menu to go to another page, there is another scenario I considered. If the user is utilizing the chat feature and wants it to remain open and visible while they explore other options, then making the background conform would be an interesting addition. that is if it's at all possible.

Also, thank you for the insight! It helped refine a few things on my end. I greatly appreciate it.

**Comment by rsanchez-wsu** [comment]
@HeadfirstRYAN I want to be careful with these sorts of purely stylistic design issues.  I would certainly not be opposed to making it an optional feature that is controlled by an application preference (perhaps in an "advanced" or "expert" mode).  Aesthetics being subjective, it is certainly conceivable that users would prefer your approach.  That would be the sort of thing that we could even make part of a usability survey of our users (once we have users, that is :-).

**Comment by Aashishkebab** [comment]
@rsanchez-wsu @HeadfirstRYAN I don't want to cause unwanted debate on this issue (heck I'm not even in the class anymore), but I do have a concern about this.

In all implementations I've seen of a hamburger menu, the content either is _under_ the hamburger menu or it is _pushed to the side_. I have never seen a hamburger menu cause content to be resized.

A concern I have with it is that performance could be degraded every time the menu is opened, as JavaFX would have to render all of the content moving around or shrinking. Furthermore, if this reorganization is animated (ideally it should be, or else the content would appear to randomly change without warning), it would cause a noticeable impact on CPU and/or GPU usage.

The only exception I've seen to this is where the hamburger menu is _always_ visible (no shadow, and no button at the top to open/close it).

If making this change, I would suggest implementing a setting in the **Application Preferences** that causes the hamburger menu to be _permanently visible_, ensuring the content redrawing only occurs once. This would almost be like changing the "theme" of the application, from a "comfortable" to a "compact" view.
In this case, when this option is enabled, the shadow and open/close buttons would have to automatically be hidden for consistency.

Here's a very shoddy version of what it could look like.

![35190524-b5345688-fe31-11e7-8188-5a17d5de0fd9](https://user-images.githubusercontent.com/25588487/35226495-7e178106-ff59-11e7-94ac-e7b765f063df.PNG)

**Comment by rsanchez-wsu** [comment]
@Aashishkebab Your concern regarding performance is quite valid.  I like your suggested compromise.  If we decide to go ahead with this, then that is the approach that we will consider as a starting point.

**Comment by HeadfirstRYAN** [comment]
@Aashishkebab Quick question. When designing the GUI for the application through the .fxml files, did you find any helpful resources online relating to the .fxml API, or really anything that proved helpful in implementing the Hamburger menu?

**Comment by Aashishkebab** [comment]
@HeadfirstRYAN I didn't really look at any resources. I already knew HTML and CSS, which helped immensely.

It is a pretty self-explanatory language, so the thing I might recommend is simply reading through the code and messing around with values to try and figure out what does what.

One thing to note is that the hamburger menu is inside a GridPane, which in turn has multiple GridPanes to add the 'separators' for the items. rowIndexes might be a little confusing because of all the different GridPanes nested.

I made quite a few changes to the hamburger menu in #94, notably e3e540d and 3a2c1b1. The latter commit is the part where I added GridPanes within the main GridPane.


Also, if you are interested in modifying the GUI using Java itself, which you would probably need to do, take a look at #106. It is quite large, but much of the content resizing would need to modify code inside of there.

0776323 sets the screen size based on resolution (though it is modified later), a4ed3ed adds everything into a ScrollPane.

**One of the most important things you'll want to look at for resizing content are 3ad5d87 and fc9ac50 and ecded61. They essentially does what you want, but based on the size of the window itself.** The second changes module size based on resolution, but you could change that to based on viewport size.
The third wraps text correctly. The first implements fluid design.

Unfortunately I don't have any 'resources' that might help, but looking at some of this code might help.

Furthermore, #117 and #174 are a Pull Requests made by another student that also modifies content sizing and stuff, that you might be able to look at.

Hopefully these help.

**Comment by HeadfirstRYAN** [comment]
@Aashishkebab I was messing around with **AnchorPane** and when I run the application and load a pre-existing profile or a new one, the values I changed to see the difference are not there. This tag is in MainMenu.fxml.

`<AnchorPane id="AnchorPane" maxHeight="-Infinity" maxWidth="-Infinity" minHeight="-Infinity" minWidth="-Infinity"
            prefHeight="750.0" prefWidth="1000.0" xmlns="http://javafx.com/javafx/8" xmlns:fx="http://javafx.com/fxml/1"
            stylesheets="@../content/stylesheet.css">`

**Comment by Aashishkebab** [comment]
@HeadfirstRYAN I am not sure I completely understand your question.

That being said, you'll want to note that when using Eclipse, if you edit any files outside of Eclipse, it does not know those changes occurred and will ignore them.

What you can try doing is right-clicking on the project and pressing "Refresh".

**Comment by HeadfirstRYAN** [comment]
Sorry, I meant to type that I changed variable values in order to see how the pane would change.
